Caching.StackExchangeRedis: add "force reconnect" pattern

mgravell opened the pull request:
# Caching.StackExchangeRedis: add "force reconnect" pattern

Adds the "force reconnect" pattern as described in [azure cache best practices](https://learn.microsoft.com/en-us/azure/azure-cache-for-redis/cache-best-practices-connection), with an implementation modified to retain the pre-existing custom connect logic

Tests forced locally (they're `[Skip]` in the source), and observed with `redis-cli monitor` to validate that key-prefix still valid.

## Description

1. add "force reconnect" pattern via OnRedisError, opt-in based on new `UseForceReconnect` option (defaults to `false`)
2. avoid thread-race problems by having Connect[Async] *return* the cache, rather than relying on field
3. avoid thread-race problems by only using a single ref field (cache), rather than cache+muxer; access muxer indirectly (it is only ever needed to shut down)
4. use more efficient key-prefix path via RedisKey.Append
5. use more efficient hash-fetch by pre-storing the hash-key arrays (removes need for RedisExtensions.cs)
6. general code tidy; null-handling, defaults, etc (no semantic change)
7. rev SE.Redis dependency to 2.6.90, which has significant connection stability improvements

Fixes #39082

mgravell committed:
1. add "force reconnect" pattern via OnRedisError
2. avoid thread-race problems by having Connect[Async] *return* the cache, rather than relying on field
3. avoid thread-race problems by only using a single ref field (cache), rather than cache+muxer; access indirectly
4. use more efficient key-prefix path via RedisKey.Append
5. use more efficient hash-fetch by pre-storing the hash-key arrays (removes need for RedisExtensions.cs)
6. general code tidy; null-handling, defaults, etc (no semantic change)

mgravell committed:
copy the time-based force reconnect logic from azure redis cache best practices

mgravell commented:
/cc @philon-msft re core reconnect eyeball

mgravell committed:
remove a couple of dead comments

NickCraver commented on src/Caching/StackExchangeRedis/src/RedisCache.cs:
```diff
@@ -93,13 +124,20 @@ internal RedisCache(IOptions<RedisCacheOptions> optionsAccessor, ILogger logger)
+            // SE.Redis allows efficient append of key-prefix scenarios, but we can help it
+            // avoid some work/allocations by forcing the key-prefix to be a byte[]; SE.Redis
+            // would do this itself anyway, using UTF8
+            _instancePrefix = Encoding.UTF8.GetBytes(instanceName);
```
Is this behavior different from `.WithPrefix()` when getting the `IDatabase`? Would tidy up a bit to switch to the built-in if not.

mgravell replied:
Semantically the same, but: as-written is more efficient in this case; if you look at `KeyspaceIsolation.WrapperBase<TInner>.ToInner(RedisKey[])`, it has to allocate a *copy* of the array each call, and `ScriptEvaluate` always uses `RedisKey[]` here. I wanted to avoid that over-and-over-and-over.

What we could do, moving forward, is add some overloads that take `ReadOnlyMemory<>` for a few key methods that currently take arrays, allowing pool leases to be used instead, but: we don't have that today.

gfoidl commented on src/Caching/StackExchangeRedis/src/RedisCache.cs:
```diff
@@ -49,20 +51,49 @@ public partial class RedisCache : IDistributedCache, IDisposable
+
+    private static DateTimeOffset ReadTimeTicks(ref long field)
+    {
+        var ticks = Interlocked.Read(ref field); // avoid torn values
```
```suggestion
        var ticks = Volatile.Read(ref field); // avoid torn values
```
? Should be enough as the exchange below has also release semantics (due the full memory barrier).

So it's cheaper at least on machine with a stronger memory model.

mgravell replied:
This is a fun topic; relevant discussion [here](https://github.com/dotnet/runtime/issues/67330#issuecomment-1083106423), but it is an interesting dilemma: the `Interlocked` approach *is documented* as being atomic; the `Volatile` approach is not so-documented, but currently *happens to be* atomic. It comes down to which guarantees we're happy to adopt. In this case, the actual implementation here is effectively copy-paste (with tweaks) from the Azure Cache repo - I'm inclined to leave it alone, honestly - since this is only in the "we're falling over" case, I'm not too concerned about a few extra fences.

halter73 replied:
From you discussion:

> The `Volatile` class, on the other hand, always performs atomic operations along with the barriers, regardless of data size.

And the docs do say:

> The [Volatile](https://learn.microsoft.com/en-us/dotnet/api/system.threading.volatile?view=net-7.0) class also provides read and write operations for some 64-bit types such as [Int64](https://learn.microsoft.com/en-us/dotnet/api/system.int64?view=net-7.0) and [Double](https://learn.microsoft.com/en-us/dotnet/api/system.double?view=net-7.0). Volatile reads and writes on such 64-bit memory are atomic even on 32-bit processors, unlike regular reads and writes.

https://learn.microsoft.com/en-us/dotnet/api/system.threading.volatile?view=net-7.0#remarks

And we do rely on this elsewhere in ASP.NET.

https://github.com/dotnet/aspnetcore/blob/63c49481c2e0462c65443a9dc68ec156a46d4d55/src/Servers/Kestrel/Core/src/Internal/Infrastructure/HeartbeatManager.cs#L21

mgravell replied:
@halter73 oh, nice; I wonder if those docs have been updated; I recall this being unclear for ages

gfoidl commented on src/Caching/StackExchangeRedis/src/RedisCache.cs:
```diff
@@ -49,20 +51,49 @@ public partial class RedisCache : IDistributedCache, IDisposable
+    private static void WriteTimeTicks(ref long field, DateTimeOffset value)
+    {
+        var ticks = value == DateTimeOffset.MinValue ? 0L : value.UtcTicks;
+        Interlocked.Exchange(ref field, ticks); // avoid torn values
```
Can `Volatile.Write` be used here?
I think yes, as it's release semantics. Would avoid the full-memory barrier introduced by `xchg` (on x64).

Atomicity is given also.

gfoidl commented on src/Caching/StackExchangeRedis/src/RedisCache.cs:
```diff
@@ -215,106 +269,118 @@ public async Task RefreshAsync(string key, CancellationToken token = default(Can
     {
-        ValidateServerFeatures();
-        TryRegisterProfiler();
+        Interlocked.Exchange(ref _lastConnectTicks, DateTimeOffset.UtcNow.UtcTicks);
```
When safe, here too. Or just call `WriteTimeTicks`.

gfoidl commented on src/Caching/StackExchangeRedis/src/RedisCache.cs:
```diff
@@ -569,4 +663,67 @@ private void CheckDisposed()
+            var firstErrorTime = ReadTimeTicks(ref _firstErrorTimeTicks);
+            if (firstErrorTime == DateTimeOffset.MinValue)
+            {
+                WriteTimeTicks(ref _firstErrorTimeTicks, utcNow);
+                WriteTimeTicks(ref _previousErrorTimeTicks, utcNow);
```
Potential races here are not a problem. Correct?
Please add a comment for this.

mgravell committed:
PR feedback

mgravell committed:
clarify timing

mgravell commented:
PR nits resolved; good feedback and discussion, thanks

halter73 commented on src/Caching/StackExchangeRedis/src/RedisCache.cs:
```diff
@@ -49,20 +51,49 @@ public partial class RedisCache : IDistributedCache, IDisposable
+    //// StackExchange.Redis will also be trying to reconnect internally,
+    //// so limit how often we recreate the ConnectionMultiplexer instance
+    //// in an attempt to reconnect
+    private readonly TimeSpan ReconnectMinInterval = TimeSpan.FromSeconds(60);
+    private readonly TimeSpan ReconnectErrorThreshold = TimeSpan.FromSeconds(30);
```
```suggestion
    // Never reconnect within 60 seconds of the last attempt to connect or reconnect.
    private readonly TimeSpan ReconnectMinInterval = TimeSpan.FromSeconds(60);
    // Only reconnect if errors have occurred for at least the last 30 seconds.
    // This count resets if there are no errors for 30 seconds
    private readonly TimeSpan ReconnectErrorThreshold = TimeSpan.FromSeconds(30);
```


Am I understanding this correctly?

Also, do we want to make `ReconnectMinInterval` and `ReconnectErrorThreshold` configurable? How about an option to disable "force reconnect" altogether?

mgravell committed:
Merge branch 'main' into marcg/forcereconnect

mgravell committed:
Merge branch 'main' into marcg/forcereconnect

mgravell committed:
Update src/Caching/StackExchangeRedis/src/RedisCache.cs

Co-authored-by: Stephen Halter <halter73@gmail.com>

mgravell commented on src/Caching/StackExchangeRedis/src/RedisCache.cs:
```diff
@@ -93,13 +124,20 @@ internal RedisCache(IOptions<RedisCacheOptions> optionsAccessor, ILogger logger)
+            // SE.Redis allows efficient append of key-prefix scenarios, but we can help it
+            // avoid some work/allocations by forcing the key-prefix to be a byte[]; SE.Redis
+            // would do this itself anyway, using UTF8
+            _instancePrefix = Encoding.UTF8.GetBytes(instanceName);
```
adding explicit conversion to reduce cognitive load

mgravell committed:
Update src/Caching/StackExchangeRedis/src/RedisCache.cs

Co-authored-by: Stephen Halter <halter73@gmail.com>

mgravell committed:
1. nits
2. add new UseForceReconnect option

mgravell committed:
rev SE.Redis to 2.6.90, which has improvements to connection stability

davidfowl commented on src/Caching/StackExchangeRedis/src/RedisCache.cs:
```diff
@@ -49,20 +51,54 @@ public partial class RedisCache : IDistributedCache, IDisposable
     private const string DataKey = "data";
+
+    // combined keys - same hash keys fetched constantly; avoid allocating an array each time
+    private static readonly RedisValue[] HashMembers_AbsoluteExpiration_SlidingExpiration_Data = new RedisValue[] { AbsoluteExpirationKey, SlidingExpirationKey, DataKey };
+    private static readonly RedisValue[] HashMembers_AbsoluteExpiration_SlidingExpiration = new RedisValue[] { AbsoluteExpirationKey, SlidingExpirationKey };
```
nit: style
```suggestion
    private static readonly RedisValue[] _hashMembersAbsoluteExpirationSlidingExpirationData = new RedisValue[] { AbsoluteExpirationKey, SlidingExpirationKey, DataKey };
    private static readonly RedisValue[] _hashMembersAbsoluteExpirationSlidingExpiration = new RedisValue[] { AbsoluteExpirationKey, SlidingExpirationKey };
```

davidfowl commented:
@BrennanConroy should we add this to the SignalR backplane as well?

Also @ReubenBond  should this be part of the redis clustering provider for orleans?

NickCraver commented:
For the other projects: if using the StackExchange.Redis client, we'd recommend upgrading to latest (2.6.x) and see if you actually hit any issues. If so, please ping ASAP and let's fix those in the core library. This entire feature _shouldn't_ be needed any more - if there are some scenarios still tossing the connection (and anything buffered, which the base library would send upon connecting up to the timeout), let's figure that out and improve for all. The workaround is inherently a bit destructive compared to the base library, and should only be a last resort.

For context: most of the need for something like this was before working with the Azure Redis cache folks to understand scenarios previously unknown and handling everything we're aware of today. TL;DR: it's mostly outdated advice, and the defaults should be doing good things today on latest library bits.

BrennanConroy commented:
> This entire feature _shouldn't_ be needed any more

So why are we doing this PR then?

mgravell commented:
@BrennanConroy because people keep finding new and exotic ways of making things go pear-shaped, and this allows us to mitigate while we investigating and resolve at the library level; which is also why it is off by default

BrennanConroy commented:
We're going to add public api for a feature that we shouldn't need to write and might be able to remove in the future? Or, if we think it's important enough to add support for it because "people keep finding new and exotic ways of making things go pear-shaped", then why not add it to our other redis libraries too?

halter73 commented on src/Caching/StackExchangeRedis/src/RedisCacheOptions.cs:
```diff
@@ -41,6 +41,11 @@ public class RedisCacheOptions : IOptions<RedisCacheOptions>
+    /// <summary>
+    /// Support the "force reconnect" pattern as described <a href="https://learn.microsoft.com/en-us/azure/azure-cache-for-redis/cache-best-practices-connection#using-forcereconnect-with-stackexchangeredis">here</a>, which can help mitigate against unstable redis connections.
+    /// </summary>
+    public bool UseForceReconnect { get; set; }
```
I agree with @BrennanConroy that if this is temporary, it might make more sense to use something like an AppContext switch rather than add public API. If we feel public API is warranted, we should take it through [api review](https://github.com/dotnet/aspnetcore/blob/main/docs/APIReviewProcess.md).

This logic is fairly complicated. I get that making this fully generic and decorating something like IConnectionMultiplexer might be too much, but could we create an internal shared-source API that participates in calling OnRedisError at the right times that can be used by our other redis-backed libraries?

mgravell replied:
App-context switch would be fine; I can tweak that if it avoids the problem

mgravell replied:
app-context switch implemented

mgravell commented:
> @BrennanConroy should we add this to the SignalR backplane as well?
> 
> Also @ReubenBond should this be part of the redis clustering provider for orleans?

what I *would* recommend is making sure you rev the lib occasionally; since me and Nick came over to MSFT and have been able to work more closely with some internal teams, we've made a bunch of stability improvements that I highly recommend

mgravell committed:
remove UseForceReconnect, preferring an app-context switch

mgravell committed:
PR feedback; move the AppContext check to RedisCacheOptions

sebastienros committed:
Merge branch 'main' into marcg/forcereconnect

mgravell committed:
- fixup NRT warnings (also uses ROM-byte support by default now)
- fixup interface delta in test harness
- make an analyzer happy about a redundant lambda

mgravell committed:
Update src/Caching/StackExchangeRedis/src/RedisCache.cs

Co-authored-by: David Fowler <davidfowl@gmail.com>

halter73 commented:
/azp run

mgravell committed:
fixup PR drift

mgravell commented:
/azp run

mgravell committed:
fix NRT

mgravell commented:
/azp run

mgravell commented:
/azp run

mgravell commented:
/azp run

mgravell commented:
/azp run
